platform/xen/rumpxendev/evtdev.c

```
@@ -0,0 +1,502 @@
+	if (err) goto out;
+
+	cons = (cons + len) & EVTDEV_RING_MASK;
+	uio_len = uio->uio_resid / BYTES_PER_PORT;
+	if (uio_len == 0) goto done;
+
+	/* ring wrapped */
+	len = prod - cons;
+	if (len > uio_len)
+		len = uio_len;
+	err = uiomove(&d->ring[cons], len * BYTES_PER_PORT, uio);
+	if (err) goto out;
+	cons = (cons + len) & EVTDEV_RING_MASK;
+
+done:
+	mutex_enter(&d->lock);
+	d->ring_cons = cons;
+	mutex_exit(&d->lock);
+out:
+	DPRINTF(("/dev/xenevt: read done, err=%d\n", err));
+	return err;
+}
+
+/*----- more exciting reading -----*/
+static int
+xenevt_dev_poll(struct file *fp, int events)
+{
+	struct xenevt_dev_data *d = fp->f_data;
+	int revents = 0;
+
+	DPRINTF(("/dev/xenevt: poll events=0x%x...\n", events));
+
+	mutex_enter(&d->lock);
+
+	/* always writable because write is used to unmask event
+	 * channel */
+	revents |= events & WBITS;
+
+	if ((events & RBITS) && (d->ring_prod != d->ring_cons))
+		revents |= events & RBITS;
+
+	/* in the case caller only interests in read but no data
+	 * available to read */
+	if (!revents && (events & RBITS))
+		selrecord(curlwp, &d->selinfo);
+
+	mutex_exit(&d->lock);
+	DPRINTF(("/dev/xenevt: poll events=0x%x done, revents=0x%x\n",
+		 events, revents));
+	return revents;
+}
+
+static int
+xenevt_dev_ioctl(struct file *fp, ulong cmd, void *data)
+{
+	struct xenevt_dev_data *d = fp->f_data;
+	int err;
+
+	switch (cmd) {
+	case IOCTL_EVTCHN_RESET:
+	{
+		mutex_enter(&d->lock);
+		d->ring_cons = d->ring_prod = 0;
+		d->flags = 0;
+		mutex_exit(&d->lock);
+		break;
+	}
+	case IOCTL_EVTCHN_BIND_VIRQ:
+	{
+		struct ioctl_evtchn_bind_virq *bind_virq = data;
+		evtchn_bind_virq_t op;
+
+		op.virq = bind_virq->virq;
+		op.vcpu = 0;
+		if ((err = minios_event_channel_op(EVTCHNOP_bind_virq, &op))) {
+			printf("IOCTL_EVTCHN_BIND_VIRQ failed: virq %d error %d\n",
+			       bind_virq->virq, err);
+			return -err;
+		}
+		bind_virq->port = op.port;
+		mutex_enter(&devevent_lock);
+		KASSERT(devevents[bind_virq->port] == NULL);
```

This should return an error instead?